src/main/scala/wasm/MiniWasmFX.scala

```diff
@@ -108,204 +63,142 @@ case class EvaluatorFX(module: ModuleInstance) {
-          eval(rest, retStack.take(funcTy.out.size) ++ restStack, frame, kont, mkont, trail, h)
-        def loop(retStack: List[Value], mkont: MCont[Ans]): Ans =
-          eval(inner, retStack.take(funcTy.inps.size), frame, restK, mkont, loop _ :: trail, h)
-        // for example, loop here doesn't change the meta-continuation at all
-        // compare with block
-        loop(inputs, mkont)
+        val escape: Cont[Ans] = (s1, t1, m1) => kont(s1.take(funcTy.out.size) ++ restStack, t1, m1)
+        def loop(retStack: List[Value], trail1: List[Cont[Ans]], mkont: MCont[Ans]): Ans =
+          evalList(inner, retStack.take(funcTy.inps.size), frame, escape, trail, mkont, (loop _ : Cont[Ans])::brTable, hs)
+        loop(inputs, trail, mkont)
       case If(ty, thn, els) =>
         val funcTy = getFuncType(ty)
         val I32V(cond) :: newStack = stack
         val inner = if (cond != 0) thn else els
         val (inputs, restStack) = newStack.splitAt(funcTy.inps.size)
-        val restK: Cont[Ans] = (retStack, mkont) =>
-          eval(rest, retStack.take(funcTy.out.size) ++ restStack, frame, kont, mkont, trail, h)
-        eval(inner, inputs, frame, restK, mkont, restK :: trail, h)
+        val escape: Cont[Ans] = (s1, t1, m1) => kont(s1.take(funcTy.out.size) ++ restStack, t1, m1)
+        evalList(inner, inputs, frame, escape, trail, mkont, escape::brTable, hs)
       case Br(label) =>
-        trail(label)(stack, mkont) // s => ().asInstanceOf[Ans]) //mkont)
+        brTable(label)(stack, trail, mkont)
       case BrIf(label) =>
         val I32V(cond) :: newStack = stack
-        if (cond != 0) trail(label)(newStack, mkont)
-        else eval(rest, newStack, frame, kont, mkont, trail, h)
+        if (cond != 0) brTable(label)(newStack, trail, mkont)
+        else kont(newStack, trail, mkont)
       case BrTable(labels, default) =>
         val I32V(cond) :: newStack = stack
         val goto = if (cond < labels.length) labels(cond) else default
-        trail(goto)(newStack, mkont)
-      case Return        => trail.last(stack, mkont)
-      case Call(f)       => evalCall(f, rest, stack, frame, kont, mkont, trail, h, false)
-      case ReturnCall(f) => evalCall(f, rest, stack, frame, kont, mkont, trail, h, true)
+        brTable(goto)(newStack, trail, mkont)
+      case Return        => brTable.last(stack, trail, mkont)
+      case Call(f)       => evalCall1(f, stack, frame, kont, trail, mkont, brTable, hs, false)
+      case ReturnCall(f) => evalCall1(f, stack, frame, kont, trail, mkont, brTable, hs, true)
       case RefFunc(f)    =>
         // TODO: RefFuncV stores an applicable function, instead of a syntactic structure
-        eval(rest, RefFuncV(f) :: stack, frame, kont, mkont, trail, h)
+        kont(RefFuncV(f) :: stack, trail, mkont)
       case CallRef(ty) =>
         val RefFuncV(f) :: newStack = stack
-        evalCall(f, rest, newStack, frame, kont, mkont, trail, h, false)
+        kont(newStack, trail, mkont)
```

Comment:
Is this right? it doesn't mention `f` at all

Reply:
oh yes, you are right, it should be a call to f